clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1108,13 +1165,39 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
     return isValid();
   }
 
+  bool handlePointerType(FieldDecl *FD, QualType FieldTy) final {
+    // TODO: Replace with a better diagnostic.
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(FD->getLocation(), diag::err_bad_kernel_param_type)
+          << FieldTy;
+    }
+    return isValid();
+  }
+
   bool handleSyclAccessorType(const CXXBaseSpecifier &BS,
                               QualType FieldTy) final {
+    // TODO: Replace with a better diagnostic.
+    if (UnionCount) {
+      IsInvalid = true;
+      Diag.Report(BS.getBeginLoc(), diag::err_bad_kernel_param_type) << FieldTy;
+      return isValid();
+    }
+
     checkAccessorType(FieldTy, BS.getBeginLoc());
     return isValid();
   }
 
   bool handleSyclAccessorType(FieldDecl *FD, QualType FieldTy) final {
+    // TODO: Replace with a better diagnostic.
+    // TODO: What other types do we need ot check?  What types other than
+    // pointer/accessor requires a decomposition?
```

Comment:
We need to check sampler and stream types. Union does not work with them.

Reply:
Cool, thanks!  I'm also considering breaking the union checking into a separate handler, so it would need to handle all of these.